### ISSUE TITLE
.stacks_out margin  problems in flex container.

### ISSUE THREAD

**Issue by atavernor** [yourhead/s3#714]
`.stacks_out` divs with `Fill Mode` set to `Fill` default to 
```
margin-left:auto;
margin-rightLauto;
```
but have no width defined.

When placed inside a container with `display:flex` these auto margins will expand to fill the available space as per the spec [here](https://www.w3.org/TR/css-flexbox-1/#annotation_auto-margins)

Note: Safari does not currently implement this spec and so the problem is only seen in Chrome & Firefox.

The anomalous result is that, by default a `.stacks_out` div will always become centered when in a flex container.

Steps to reproduce:

- Place a header stack on the page. (It should be aligned left)
- Leave its fill mode set to `Fill`
- Preview the page in Chrome.
- In the inspector, apply `display:flex` to its parent (i.e. `.stacks_top>div`) 

Result:

The header becomes aligned center on the page instead of left aligned.

Obviously if the `Fill mode` is changed to `Flexible` then a `width:100%` is applied which will overcome the flex growth of the margins and the stack behaves as expected.

Possible Solutions:
Are the instructions `margin:left:auto` and `margin-right:auto` really serving any purpose when the `Fill Mode` is set to `Fill` ?  If not then removing them would solve the problem.

The possibly safer alternative would seem to be to add a `width:100%` to `,stacks_out` when the `Fill Mode` is set to `Fill`. This would appear to have no deleterious effects and seems a logical condition.

**Comment by yourhead** [comment]
this is interesting -- but doesn't seem like a real problem (for now) -- flex box is still too new to be used within stacks since stacks guarantees cross-browser compatibility with 95% of the browsers in use today -- something that flex box probably won't have for a year or more.

your method of reproducing this bug is to tweak the stacks_top div to be a `display:flex`
the top div is a standard stacks container div. all stacks container divs are guaranteed to be display:block -- we do this not only for cross-browser compatibility -- but also to keep-it-simple for novice users unfamiliar with the complexities of web layout. so this situation is unlikely to change -- irregardless of browsers -- so long as most of the users of stacks are novice ones (hint: they are by a wide margin -- so this is unlikely to ever change)

we will totally add flex-box to stacks (because it's AWESOME!!! 😝 ) -- when it because more consistently implemented -- and older devices that implement it poorly or not at all become a smaller portion of the browsing public. when we do we'll insure that it works correctly with all the various modes of stacks.

if you think i'm missing the point here -- or that flex box is in use within stacks in some way that i'm not seeing -- please let me know.  i'm open to being convinced.  😃 

but tweaking the browser css is probably the wrong way to go about it. show me how it applies to a real-world situation. use the built in stacks -- or a well made 3rd party stack. provide simple steps (not tweaking CSS!) to reproduce.

thanks,
isaiah

**Comment by atavernor** [comment]
Flex box is being used in many stacks already both mine and several other developers. 

It is supported by all non end of life browsers (even IE10 works just fine for almost all of it) 

[http://caniuse.com/#feat=flexbox](http://caniuse.com/#feat=flexbox)

>your method of reproducing this bug is to tweak the stacks_top div

That was just the quickest way to demonstrate the problem. Any div that is `display:flex` with a stacks dropzone in it will obviously show the behavior. 

The reason I mentioned it was because users reported it and the only solution is to use 

`.myDiv>.stacks_out {width :100%}`

which works fine but is a course solution and could possibly mess up content where people had set flexible widths using the stacks inbuilt layout controls

**Comment by atavernor** [comment]
The most obvious real world examples are grids and columns. When using flex box these will inevitably have `.stacks_out` dropped directly into a `display:flex` div.  

Frameworks are already using flex box primarily and dropping the float grid out in the wild (although floats are still supported they are becoming very much advised against and will become legacy). At least one (Kube) is already in use in RapidWeaver. Future versions of Foundation will use flex box and it is now being accepted as the norm until CSS Grid gains wide browser support.

I don't thing that width:100% on wrappers that are set to `Fill` is too great a leap of faith to potentially avoid a lot of problems. Maybe I am missing something there but it seems like a win - win.

**Comment by yourhead** [comment]
i'll investigate

**Comment by atavernor** [comment]
The alternative of course would just be to provide class names on the `.stacks_out` div indicating the state of the Fill Mode control. In this way no potentially disruptive changes would be necessary to the Stacks core CSS and the onus would be on developers using Flex Box to handle the content appropriately.

For example, if my `display:flex` div sees `>.stacks_out.stacks-fill` then it can safely remove the auto margins (preferable) or apply a `width:100%`

**Comment by yourhead** [comment]
started a branch for this to test out some implementations, consequences.  this is my notes.  @atavernor -- this isn't in response to any of your details -- i just wanted to make sure and capture my thoughts for this branch.

#### bug
`Fill Mode: Fill` implies a width of 100%. the current implementation does not produce that behavior inside of containers that use `display:flex`.

#### proposed implementations
 - Ensure an explicit width:100% on all divs with this implied behavior. This seems like the *right* fix.
 - Just adding a class seems like we're "coding to the bug" : i.e. fixing the symptom -- or in that case -- adding a feature that lets the stack-dev work around the symptom. nope nope nope.


#### requirements
Add: `width:100%` to:

- publish mode: out div 
- edit mode: 
  - container wrapper div
  - mez div on bare nodes

#### risk
low, but not non-zero
there are many stacks that currently override the the css of the out-div and/or the container wrapper in edit mode. modifying these could potentially impact stacks that are applying their own width and/or max width. i see no means of testing for real impact in all stacks.

#### mitigations
unfortunately it does not seem possible to introduce an opt-in or compatibility mode switch for this change. to be effective the behavior has to change universally on every stack that could be placed into a flex container.

#### when
since there is a non-zero risk of breaking backward compatibility in published sites -- we have to introduce this change in a very tactical way. introducing a breaking change to v3.5.0 is a no-go. i've already gone public with "no breaking changes".  introducing it at v4.0 could have business implications.

i think we'll just need to find a "sneaky" time to introduce this -- sort of toss it out in a 3.5.x beta -- and see if it sticks.  shouldn't wait too long -- since it would be nice to have a good chunk of testers looking at it -- but don't want it to be right on the heals of 3.5 -- we want the uptake of 3.5 to be as close to 100% as possible.

**Comment by atavernor** [comment]
Yes, that all makes sense, but let me offer a different perspective to think about.

The margins are the cause of the problem. In fill mode they do not serve a purpose, the out div is inherently 100% and we are introducing left and right auto margins so far as I can tell for convenience further down the line i.e. when flexible, align right is selected, the right margin simply needs setting to 0 as the left is already set to auto.

It is these margins that are the problem in a flex container as they will grow (or shrink).

In this respect, the width is not the problem and setting the width to 100% is a relatively crude fix to overcome the redundant (in non flex situations) margins.

It therefore seems sensible to look at solving the root of the problem by removing the auto left and right margins rather than glossing over it somewhat by the imposition of a 100% width.

Risks:

Unlike the non-present width, developers who need to override the margins on the out div already need to use `!important` 

There is of course the risk that the assumption is made that the non modified margin is already at auto and so problems could of course still arise.

The risks are therefore lower but still present and so arguments about when to implement concur with your statements about the width solution.

**Comment by yourhead** [comment]
agreed

**Comment by yourhead** [comment]
see: https://github.com/yourhead/s4/issues/26
